### ISSUE TITLE
Unexpected warnings with Standardize in ChainedOutcomeTransform

### ISSUE THREAD

**Issue by Balandat** [pytorch/botorch#2812]
### Discussed in https://github.com/pytorch/botorch/discussions/2811

<div type='discussions-op-text'>

<sup>Originally posted by **samuelkim16** April  8, 2025</sup>
I'd like to use multiple outcome transforms. A snippet of my code looks like this:

```python
import botorch.models.transforms as transforms

outcome_transform = transforms.outcome.ChainedOutcomeTransform(
    tf1=transforms.outcome.Log(),
    tf2=transforms.outcome.Standardize(m=1),
)

model = SingleTaskGP(train_X, train_Y,
    input_transform=input_transform,
    outcome_transform=outcome_transform, 
)
```

However, during my optimization run, I will regularly get warnings in each iteration like the following:
```
WARNING:py.warnings:InputDataWarning: Data (outcome observations) is not standardized (std = tensor([31.5549], dtype=torch.float64), mean = tensor([-1.2118], dtype=torch.float64)).Please consider scaling the input to zero mean and unit variance.
 C:\Users\username\AppData\Roaming\Python\Python310\site-packages\botorch\models\utils\assorted.py: 267
```

Am I chaining the outcome transforms incorrectly? The warning does not give a full stack trace so it is difficult for me to troubleshoot. I do not get any errors when using Standardize alone (without ChainedOutcomeTransform) or when just initializing the model.</div>

**Comment by Balandat** [comment]
See this comment: https://github.com/pytorch/botorch/discussions/2811#discussioncomment-12779398

**Comment by samuelkim16** [comment]
Apologies, I think I figured out the issue.

When refactoring my code to use a `ChainedOutcomeTransform`, I had moved the transform outside of the optimization loop. Thus, it was standardizing the data the first time I trained the GP model, but raised warnings on every subsequent iteration. Below is a MWE that will bring up the error.

```python
# BoTorch imports
import torch
import botorch.models.transforms as transforms
from botorch.models.gp_regression_fidelity import SingleTaskGP
from gpytorch.mlls.exact_marginal_log_likelihood import ExactMarginalLogLikelihood
from botorch import fit_gpytorch_mll
from botorch.acquisition import qLogExpectedImprovement
from botorch.optim import optimize_acqf

# Set up problem and initial data
problem_dim = 2
train_X = torch.rand(10, problem_dim, dtype=torch.double)
train_Y = torch.linalg.norm(train_X - 0.5, dim=-1, keepdim=True)
bounds = torch.stack([torch.zeros(problem_dim), torch.ones(problem_dim)])

outcome_transform = transforms.outcome.ChainedOutcomeTransform(
    tf1=transforms.outcome.Log(),
    tf2=transforms.outcome.Standardize(m=1),
)

for i in range(5):
    print(f"Iteration: {i+1}")

    # Train model
    model = SingleTaskGP(
        train_X=train_X, train_Y=train_Y,
        input_transform=transforms.input.Normalize(d=problem_dim),
        outcome_transform=outcome_transform, 
    )
    mll = ExactMarginalLogLikelihood(model.likelihood, model)
    fit_gpytorch_mll(mll)

    # Collect next data points
    logEI = qLogExpectedImprovement(model=model, best_f=train_Y.max())
    X_new, _ = optimize_acqf(
        logEI, bounds=bounds, q=2, num_restarts=5, raw_samples=20,
    )
    Y_new = torch.linalg.norm(X_new - 0.5, dim=-1, keepdim=True)

    print(f"X_new: \n{X_new}")
    print(f"Y_new: \n{Y_new}")

    # Add new data points to dataset
    train_X = torch.cat([train_X, X_new])
    train_Y = torch.cat([train_Y, Y_new])
```

Moving the outcome transform inside the loop (such that it re-initializes in each iteration) gets rid of the error. It was not clear to me reading the documentation that once trained, the Standardize object switches to (and stays in) eval mode. I had assumed that it goes to train mode when training the GP model.

**Comment by Balandat** [comment]
Thanks - I think I know what's going on. The outcome transform is applied to the targets in the constructor of the `SingleTaskGP` here: https://github.com/pytorch/botorch/blob/main/botorch/models/gp_regression.py#L165-L168

If it's not already in `train` mode, then it's applied to the targets without updating the scaling. Simply adding a `outcome_transform.train()` before the application of it fixes the issue. There are probably a number of other call sites and possibly interactions with other things, so let me make sure this change doesn't cause any unforeseen side effects. Will put up a PR with a fix.

**Comment by Balandat** [comment]
Interestingly, we are doing this in at least one place: https://github.com/pytorch/botorch/blob/main/botorch/models/fully_bayesian_multitask.py#L249. Whoever made that change clearly has no idea what they're doing (hint: it was me).
